cargo test fails in rust 1.64.0

The output for `cargo test`

```
$ cargo test
    Finished test [unoptimized + debuginfo] target(s) in 0.25s
     Running unittests src/hoice.rs (target/debug/deps/hoice-8c0ebc4b1caf9f8b)

running 40 tests
test dtyp::dtyp_write_dec ... ok
...(omitted)....
test src/common/macros.rs - common::macros::try_val (line 382) ... FAILED
...(omitted)...
---- src/common/macros.rs - common::macros::try_val (line 382) stdout ----
error: cannot find macro `try_val` in this scope
  --> src/common/macros.rs:390:7
   |
11 |   Ok( try_val!{ bool val } )
   |       ^^^^^^^

error: cannot find macro `try_val` in this scope
 --> src/common/macros.rs:387:7
  |
8 |   Ok( try_val!{ int val } )
  |       ^^^^^^^

error[E0412]: cannot find type `Res` in this scope
 --> src/common/macros.rs:386:21
  |
7 | fn int(val: Val) -> Res<Val> {
  |                     ^^^ not found in this scope
  |
help: consider importing this type alias
  |
2 | use hoice::errors::Res;
  |

error[E0412]: cannot find type `Res` in this scope
  --> src/common/macros.rs:389:22
   |
10 | fn bool(val: Val) -> Res<Val> {
   |                      ^^^ not found in this scope
   |
help: consider importing this type alias
   |
2  | use hoice::errors::Res;
   |

error[E0603]: type alias `Val` is private
  --> src/common/macros.rs:383:18
   |
4  | use hoice::term::Val ;
   |                  ^^^ private type alias
   |
note: the type alias `Val` is defined here
  --> /home/katsura/github.com/hopv/hoice/src/term.rs:60:5
   |
60 | use crate::common::*;
   |     ^^^^^^^^^^^^^^^^

error: unused import: `hoice::errors`
 --> src/common/macros.rs:384:5
  |
5 | use hoice::errors ;
  |     ^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/common/macros.rs:380:9
  |
1 | #![deny(warnings)]
  |         ^^^^^^^^
  = note: `#[deny(unused_imports)]` implied by `#[deny(warnings)]`

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0412, E0603.
For more information about an error, try `rustc --explain E0412`.
Couldn't compile the test.

failures:
    src/common/macros.rs - common::macros::try_val (line 382)

test result: FAILED. 215 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 53.01s
```

It looks like the doctest for `common::macros::try_val` is actually broken, but in the past stable cargo did not complain about it because it ignored that test. In fact, if I switch my toolchain version to 1.60.0, then `cargo test` successfully passes since it seems to ignore this broken test (well, presumably, cargo in 1.60.0 was broken with some reason 🤷 ).

I'm not sure about the implementation of `hoice`, so I cannot fix this problem myself, but the test should be fixed so that CI won't complain anymore about this.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -9,7 +9,7 @@ readme = "README.md"
 categories = ["science"]
 keywords = ["machine learning", "verification", "proof"]
 license = "MIT"
-edition = "2018"
+edition = "2021"
 
 [[bin]]
 name = "hoice"
@@ -32,16 +32,16 @@ panic = 'unwind'
 bench = [ ]
 
 [dependencies]
-libc = "*"
-lazy_static = "*"
-clap = { version = "*", features = ["cargo"] }
-hashconsing = "^1"
-error-chain = "*"
-ansi_term = "*"
+libc = "^0.2"
+lazy_static = "^1.4"
+clap = { version = "^3.1", features = ["cargo"] }
+hashconsing = "^1.5"
+error-chain = "^0.12"
+ansi_term = "^0.12"
 rsmt2 = "^0.16"
-num = "*"
+num = "^0.4"
 mylib = { git = "https://github.com/AdrienChampion/mylib" }
-either = "*"
-rand = "*"
-rand_xorshift = "*"
-atty = "*"
+either = "^1.8"
+rand = "^0.8"
+rand_xorshift = "^0.3"
+atty = "^0.2"

--- a/src/check.rs
+++ b/src/check.rs
@@ -9,7 +9,7 @@
 //! is no real notion of term here.
 //!
 //! [hoice]: https://github.com/hopv/hoice (hoice github repository)
-//! [smt]: http://smtlib.cs.uiowa.edu/ (SMT-LIB website)
+//! [smt2]: http://smtlib.cs.uiowa.edu/ (SMT-LIB website)
 
 use crate::{
     common::{conf, ColorExt, HashMap, Instance, Read, Solver},

--- a/src/common/config.rs
+++ b/src/common/config.rs
@@ -12,7 +12,7 @@ use rsmt2::SmtConf as SolverConf;
 
 use crate::{common::mk_dir, errors::*, instance::Instance};
 
-/// Creates a function adding arguments to a `::clap::App`.
+/// Creates a function adding arguments to a `clap::App`.
 macro_rules! app_fun {
     // Internal rules.
     (@app $app:expr, $order:expr =>) => ($app);
@@ -90,7 +90,7 @@ macro_rules! make_conf {
             long_help $long_help:expr,
             $($tail:tt)*
         } {
-            |$mtch:pat| $field_do:expr
+            |$mtch:ident| $field_do:expr
         }
     )*}
 
@@ -137,9 +137,9 @@ macro_rules! make_conf {
 }
 
 /// Clap `App` with static lifetimes.
-pub type App = ::clap::Command<'static>;
+pub type App = clap::Command<'static>;
 /// Clap `ArgMatches` with static lifetime.
-pub type Matches = ::clap::ArgMatches;
+pub type Matches = clap::ArgMatches;
 
 /// Functions all sub-configurations must have.
 pub trait SubConf {

--- a/src/common/macros.rs
+++ b/src/common/macros.rs
@@ -3,7 +3,7 @@
 /// If the input is an error, prints it SMT-LIB-style and panics.
 #[macro_export]
 macro_rules! expect {
-  ($e:expr => |$err:pat| $($action:tt)*) => (
+  ($e:expr => |$err:ident| $($action:tt)*) => (
     match $e {
       Ok(res) => res,
       Err($err) => {
@@ -376,33 +376,7 @@ macro_rules! msg {
     };
 }
 
-/// Yields the value if the type (int or bool) matches, otherwise
-/// `return`s `Ok(Val::N)`.
-///
-/// ```rust
-/// use hoice::term::Val ;
-/// use hoice::errors ;
-///
-/// fn int(val: Val) -> Res<Val> {
-///   Ok( try_val!{ int val } )
-/// }
-/// fn bool(val: Val) -> Res<Val> {
-///   Ok( try_val!{ bool val } )
-/// }
-///
-/// let none = Val::N ;
-///
-/// let val: Val = 7.into() ;
-/// assert_eq!{ int( val.clone() ), val }
-/// assert_eq!{ bool( val.clone() ), none }
-///
-/// let val: Val = false.into() ;
-/// assert_eq!{ int( val.clone() ), none }
-/// assert_eq!{ bool( val.clone() ), val }
-///
-/// assert_eq!{ int( none.clone() ), none }
-/// assert_eq!{ bool( none.clone() ), none }
-/// ```
+/// Yields the value if the type (int or bool) matches, otherwise `return`s `Ok(Val::N)`.
 macro_rules! try_val {
     (int $e:expr) => {
         if let Some(i) = $e.to_int()? {

--- a/src/term.rs
+++ b/src/term.rs
@@ -70,12 +70,7 @@ mod tterms;
 pub mod typ;
 mod zip;
 
-pub use self::bindings::Bindings;
-pub use self::factory::*;
-pub use self::leaf_iter::LeafIter;
-pub use self::op::*;
-pub use self::tterms::*;
-pub use self::typ::Typ;
+pub use self::{bindings::Bindings, factory::*, leaf_iter::LeafIter, op::*, tterms::*, typ::Typ};
 
 #[cfg(test)]
 mod test;

